File: learning-spring-java/src/main/java/com/ironbird/learningspring/business/ReservationService.java
package com.ironbird.learningspring.business;

import com.ironbird.learningspring.data.*;
import org.springframework.stereotype.Service;

import java.util.*;

@Service
public class ReservationService {
    private final RoomRepository roomRepository;
    private final GuestRepository guestRepository;
    private final ReservationRepository reservationRepository;

    public ReservationService(RoomRepository roomRepository, GuestRepository guestRepository, ReservationRepository reservationRepository) {
        this.roomRepository = roomRepository;
        this.guestRepository = guestRepository;
        this.reservationRepository = reservationRepository;
    }

    public List<RoomReservation> getRoomReservationsForDate(Date date) {
        Iterable<Room> rooms = this.roomRepository.findAll();
        Map<Long, RoomReservation> roomReservationMap = new HashMap();
        rooms.forEach(room -> {
            RoomReservation roomReservation = new RoomReservation();
            roomReservation.setRoomId(room.getRoomId());
            roomReservation.setRoomName(room.getName());
            roomReservation.setRoomNumber(room.getRoomNumber());
            roomReservationMap.put(room.getRoomId(), roomReservation);
        });

        Iterable<Reservation> reservations = this.reservationRepository.findReservationByReservationDate(new java.sql.Date(date.getTime()));
        reservations.forEach(reservation -> {
            RoomReservation roomReservation = roomReservationMap.get(reservation.getRoom().getRoomId());
            roomReservation.setDate(date);
            Guest guest = this.guestRepository.findById(reservation.getReservationId()).get();
            roomReservation.setFirstName(guest.getFirstName());
            roomReservation.setLastName(guest.getLastName());
            roomReservation.setGuestId(guest.getGuestId());
        });

        List<RoomReservation> roomReservations = new ArrayList<>();
        for (Long id : roomReservationMap.keySet()) {
            roomReservations.add(roomReservationMap.get(id));
        }
        roomReservations.sort(new Comparator<RoomReservation>() {
            @Override
            public int compare(RoomReservation o1, RoomReservation o2) {
                if (o1.getRoomName().equals(o2.getRoomName())) {
                    return o1.getRoomNumber().compareTo(o2.getRoomNumber());
                }
                return o1.getRoomName().compareTo(o2.getRoomName());
            }
        });
        return roomReservations;
    }


    public List<Guest> getHotelGuests() {
        Iterable<Guest> guests = guestRepository.findAll();
        List<Guest> guestsResult = new ArrayList<>();
        guests.forEach(guestsResult::add);

        guestsResult.sort(new Comparator<Guest>() {
            @Override
            public int compare(Guest o1, Guest o2) {
                if (o1.getLastName() == null || o1.getLastName().equals(o2.getLastName())) {
                    return o1.getFirstName().compareTo(o2.getFirstName()); // TODO not catering for o1.firstName being null
                }
                return o1.getLastName().compareTo(o2.getLastName());
            }
        });

        return guestsResult;
    }


    public void addGuest(Guest guest) {
        if (guest == null) {
            throw new RuntimeException("Cannot add a null Guest");
        }

        this.guestRepository.save(guest);
    }

    public List<Room> getRooms() {
        Iterable<Room> rooms = roomRepository.findAll();
        List<Room> roomsResult = new ArrayList<>();
        rooms.forEach(roomsResult::add);

        roomsResult.sort(new Comparator<Room>() {
            @Override
            public int compare(Room o1, Room o2) {
                return o1.getRoomNumber().compareTo(o2.getRoomNumber());
            }
        });

        return roomsResult;
    }
}

File: learning-spring-kotlin/src/main/kotlin/com/ironbird/learningspring/business/ReservationService.kt
package com.ironbird.learningspring.business

import com.ironbird.learningspring.data.*
import org.springframework.stereotype.Service
import java.util.*


@Service
class ReservationService(
    private val reservationRepository: ReservationRepository,
    private val roomRepository: RoomRepository,
    private val guestRepository: GuestRepository
) {

    fun getRoomReservationsForDate(date: Date): List<RoomReservation> {
        val rooms = roomRepository.findAll()

        val roomReservationMap = rooms.associate { room ->
            room.roomId to RoomReservation(
                roomId = room.roomId,
                roomName = room.name,
                roomNumber = room.roomNumber
            )
        }

        val reservations = reservationRepository.findByReservationDate(java.sql.Date(date.time))
        reservations?.forEach {
            val roomReservation = roomReservationMap[it?.room?.roomId]
            roomReservation?.date = it?.reservationDate
            roomReservation?.guestId = it?.guest?.guestId
            roomReservation?.firstName = it?.guest?.firstName
            roomReservation?.lastName = it?.guest?.lastName
        }

        val reservationsForDate = roomReservationMap.flatMap { (_, roomReservation) -> listOf(roomReservation) }


        return reservationsForDate.sortedBy { it.roomNumber }
    }

    fun getHotelGuests(): List<Guest> {
        val guests = guestRepository.findAll().toList()

        return guests.sortedWith { g1, g2 ->
            if (g1.lastName == null || g1.lastName == g2.lastName) {
                g1.firstName?.compareTo(g2.firstName ?: "") ?: -1
            } else {
                g1.lastName.compareTo(g2.lastName ?: "")
            }
        }
    }

    fun addHotelGuest(guest: Guest?) {
        if (guest == null) {
            throw RuntimeException("Cannot add a null Guest")
        }
        guestRepository.save(guest)
    }


    fun getRooms(): List<Room> {
        val rooms = roomRepository.findAll().toList()

        return rooms.sortedBy { room -> room.roomNumber }
    }
}